sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBoundAttribute.scala

```diff
@@ -17,49 +17,73 @@
 package com.nvidia.spark.rapids
 
 import org.apache.spark.internal.Logging
-import org.apache.spark.sql.catalyst.expressions.AttributeSeq
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, AttributeSeq, Expression, SortOrder}
 import org.apache.spark.sql.types.DataType
 import org.apache.spark.sql.vectorized.ColumnarBatch
 
 object GpuBindReferences extends Logging {
 
+  private[this] def postBindCheck[A <: Expression](base: A): Unit = {
+    base.foreach { expr =>
+      // The condition is needed to have it match what transform
+      // looks at, otherwise we can check things that would not be modified.
+      if (expr.containsChild.nonEmpty) {
+        expr match {
+          case _: GpuExpression =>
+          case _: SortOrder =>
+          case other =>
+            throw new IllegalArgumentException(
+              s"Bound an expression that shouldn't be here ${other.getClass}")
+        }
+      }
+    }
+  }
+
   // Mostly copied from BoundAttribute.scala so we can do columnar processing
-  def bindReference[A <: GpuExpression](
+  private[this] def bindRefInternal[A <: Expression, R <: Expression](
       expression: A,
-      input: AttributeSeq,
-      allowFailures: Boolean = false): A = {
-    expression.transform { case a: GpuAttributeReference =>
-      val ordinal = input.indexOf(a.exprId)
-      if (ordinal == -1) {
-        if (allowFailures) {
-          a
-        } else {
+      input: AttributeSeq): R = {
+    val ret = expression.transform {
+      case a: AttributeReference =>
```

Comment:
transform walks through the tree and if the supplied pattern matches something in the tree it will be replaced with what is returned.  So the goal of the code is to replace all `AttributeReference` instances with `GpuBoundReference` instances.  Everything else passes through.